pkg/prometheus/agent_http.go

```diff
@@ -0,0 +1,231 @@
+		return nil, err
+	}
+
+	return &configapi.GetConfigurationResponse{Value: cfg}, nil
+}
+
+// PutConfiguration creates or updates a named configuration. Completely
+// overrides the previous configuration if it exists.
+func (a *Agent) PutConfiguration(r *http.Request) (interface{}, error) {
+	ctype, err := configapi.ContentTypeFromRequest(r)
+	if err != nil {
+		return nil, err
+	}
+
+	inst, err := UnmarshalInstanceConfig(r.Body, ctype)
+	if err != nil {
+		return nil, err
+	}
+	inst.Name = getConfigName(r)
+
+	var newConfig bool
+	err = a.kv.CAS(r.Context(), inst.Name, func(in interface{}) (out interface{}, retry bool, err error) {
+		// The configuration is new if there's no previous value from the CAS
+		newConfig = (in == nil)
+		return inst, false, nil
+	})
+
+	if err == nil {
+		if newConfig {
+			totalCreatedConfigs.Inc()
+		} else {
+			totalUpdatedConfigs.Inc()
+		}
+	}
+
+	return nil, err
+}
+
+// DeleteConfiguration deletes an existing named configuration.
+func (a *Agent) DeleteConfiguration(r *http.Request) (interface{}, error) {
+	configKey := getConfigName(r)
+
+	v, err := a.kv.Get(r.Context(), configKey)
+	if err != nil {
+		// Silently ignore the error; Get is just used for validation since Delete
+		// will never return an error if the key doesn't exist. We'll log it anyway
+		// but the user will be left unaware of there being a problem here.
+		level.Error(a.logger).Log("msg", "error validating key existence for deletion", "err", err)
+	} else if v == nil {
+		// But if the object doesn't exist, there's nothing to delete.
+		return nil, &httpError{
+			StatusCode: http.StatusBadRequest,
+			Err:        fmt.Errorf("configuration %s does not exist", configKey),
+		}
+	}
+
+	err = a.kv.Delete(r.Context(), configKey)
+	if err != nil {
+		level.Error(a.logger).Log("msg", "error deleting configuration from kv store", "err", err)
+		return nil, fmt.Errorf("error deleting configuration: %w", err)
+	}
+
+	totalDeletedConfigs.Inc()
+	return nil, err
+}
+
+type httpError struct {
+	StatusCode int
+	Err        error
+}
+
+func (e httpError) Error() string { return e.Err.Error() }
+
+// getConfigName uses gorilla/mux's route variables to extract the
+// "name" variable. If not found, getConfigName will panic.
+func getConfigName(r *http.Request) string {
+	vars := mux.Vars(r)
+	name, ok := vars["name"]
+	if !ok {
+		panic("name not found in mux vars - the mux route is misconfigured")
+	}
+	return name
+}
+
+// UnmarshalInstanceConfig unmarshals an instance config from a reader
+// based on a provided content type.
+func UnmarshalInstanceConfig(r io.Reader, ctype configapi.ContentType) (*InstanceConfig, error) {
+	var (
+		cfg InstanceConfig
+		err error
+	)
+
+	switch ctype {
+	case configapi.ContentTypeJSON:
+		err = json.NewDecoder(r).Decode(&cfg)
+	case configapi.ContentTypeYAML:
+		err = yaml.NewDecoder(r).Decode(&cfg)
+	default:
+		panic(fmt.Sprintf("unhandled content type %s", ctype))
```

Comment:
Should we be panicking for an unknown content type? I actually don't think this code is reachable given you're defending against this scenario when getting the content type.